### PR TITLE
Upgrade React from 15 to 16

### LINKED ISSUE #564: Upgrade React from 15.x to 16.x
This would give us access to https://reactjs.org/docs/fragments.html for example (which would be very useful to properly fix #562), and keep up to date with the main dependency of the project.

Regarding [upgrading React itself](https://reactjs.org/blog/2017/09/26/react-v16.0.html#upgrading) it should be rather straightforward, but we might encounter issues with dependencies, and deprecations (like `react-addons-test-utils` which [needs to be migrated](https://reactjs.org/blog/2017/04/07/react-v15.5.0.html#discontinuing-support-for-react-addons)).

### PR DESCRIPTION
**Pull request by magopian**
Fixes #564

- [X] Update package versions
- [X] Migrate old react-router's IndexRoute and IndexRedirect
- [x] Migrate "common" components route (Notifications, Sidebar)
- [X] Migrate HomePage route (including AuthForm)
- [x] Migrate plugins dynamic routes
- [x] Migrate breadcrumbs
- [x] Migrate Buckets route
- [x] Migrate groups route
- [x] Migrate Collections route
- [x] Migrate Records route
- [x] Migrate 404 route
- [x] `location.query` [doesn't exist anymore in react-router v4](https://codeburst.io/react-router-v4-unofficial-migration-guide-5a370b8905a#27d7)
- [x] Update the react-redux integration by replacing react-router-redux with [connected-react-router](https://github.com/supasate/connected-react-router)
- [x] There's no `onEnter` nor `onUpdate` on `Route`s [anymore](https://codeburst.io/react-router-v4-unofficial-migration-guide-5a370b8905a#e85f)
- [x] Use a `SmartRoute` HOC (see [CrumbRoute](https://github.com/svenanders/react-breadcrumbs/blob/b71986a33284188bd3f83d47c34eb8a60c00d35b/demo/src/crumb-route.jsx#L16)) that
  - [x] calls `routeActions.routeUpdated(params, location)` with the `Route` params and location
  - [x] adds a `Breadcrumb` component (to fix the breadcrumbs)
- [x] fix the `Sidebar` so it properly highlights the current bucket and collection when needed
- [x] fix tests
- [x] fix flow
- [x] fix issue with [bucket history](http://localhost:3000/#/buckets/testbucket/history) and [group history](http://localhost:3000/#/buckets/testbucket/groups/testgroup/history) and [collections history](http://localhost:3000/#/buckets/testbucket/collections/testcollection/history) and [records history](http://localhost:3000/#/buckets/testbucket/collections/testcollection/records/<:rid>/history) when refreshing the page (`TypeError: routing.locationBeforeTransitions is null`)
- [x] fix issue with records being listed several times in the collection's record list

### CODE REVIEW

**Comment by leplatrem** [comment]
I tested locally with the signoff plugin etc. and it works great :)

File: src/components/AdminLink.js [hunk 1]
```diff
@@ -4,21 +4,26 @@ import type { RouteParams } from "../types";
...
   children?: React.Node,
+  query?: Object,
 };
 
 export default class AdminLink extends PureComponent<Props> {
   render() {
     const { children, name, params, ...linkProps } = this.props;
+    const toUrl = url(name, params);
+    const to =
+      (linkProps.query && `${toUrl}?${stringify(linkProps.query)}`) || toUrl;
```
Comment (leplatrem) [inline src/components/AdminLink.js]: ``const to = linkProps.query ? `${toUrl}?${stringify(linkProps.query)}` : toUrl``

File: src/components/AdminLink.js [hunk 2]
```diff
@@ -4,21 +4,26 @@ import type { RouteParams } from "../types";
...
     return (
-      <Link {...linkProps} to={url(name, params)}>
+      <Link {...linkProps} to={to}>
```
Comment (leplatrem) [inline src/components/AdminLink.js]: Couldn't we have written ` linkProps = { ...linkProps, to}` above and just pass `{ ...linkProps }` here?

File: src/components/bucket/BucketAttributes.js [hunk 3]
```diff
@@ -35,14 +35,14 @@ export default class BucketAttributes extends PureComponent<Props> {
   };
 
   onSubmit = (formData: BucketData) => {
-    const { params, updateBucket } = this.props;
-    const { bid } = params;
+    const { match, updateBucket } = this.props;
+    const { bid } = match.params;
```
Comment (leplatrem) [inline src/components/bucket/BucketAttributes.js]: `const { params: { bid } } = match`

File: src/components/bucket/BucketAttributes.js [hunk 4]
```diff
@@ -35,14 +35,14 @@ export default class BucketAttributes extends PureComponent<Props> {
...
     updateBucket(bid, { data: formData });
   };
 
   render() {
-    const { params, session, capabilities, bucket } = this.props;
-    const { bid } = params;
+    const { match, session, capabilities, bucket } = this.props;
+    const { bid } = match.params;
```
Comment (leplatrem) [inline src/components/bucket/BucketAttributes.js]: `const { params: { bid } } = match`

File: src/components/collection/CollectionHistory.js [hunk 5]
```diff
@@ -18,23 +20,47 @@ type Props = {
...
+    cid: string,
+    filters: HistoryFilters
+  ) => void,
   listCollectionNextHistory: () => void,
   notifyError: (message: string, error: ?Error) => void,
 };
 
+export const onCollectionHistoryEnter = (props: Props) => {
+  const { listCollectionHistory, match, session, location } = props;
+  const { bid, cid } = match.params;
```
Comment (leplatrem) [inline src/components/collection/CollectionHistory.js]: `const {params: { bid, cid }} = match;`

File: src/components/group/GroupAttributes.js [hunk 6]
```diff
@@ -26,22 +26,22 @@ type Props = {
 
 export default class GroupAttributes extends PureComponent<Props> {
   onSubmit = (formData: GroupData) => {
-    const { params, updateGroup } = this.props;
-    const { bid, gid } = params;
+    const { match, updateGroup } = this.props;
+    const { bid, gid } = match.params;
```
Comment (leplatrem) [inline src/components/group/GroupAttributes.js]: `const { params: { bid, gid }} = match;`

File: src/index.js [hunk 7]
```diff
@@ -56,16 +54,11 @@ export default class KintoAdmin extends Component<Props> {
...
-      store.dispatch(routeActions.routeUpdated(params, location));
-    }
-
     return (
       <Provider store={store}>
-        <Router history={hashHistory} onUpdate={onRouteUpdate}>
-          {getRoutes(this.store, registerPlugins)}
-        </Router>
+        <ConnectedRouter history={hashHistory}>
+          <Switch>{getRoutes(this.store, registerPlugins)}</Switch>
```
Comment (leplatrem) [inline src/index.js]: Shouldn't the `<Switch>` be in `getRoutes()`? (genuine question, maybe not)

File: src/components/AdminLink.js (hunk 1 above)
Comment (magopian) [reply src/components/AdminLink.js]: Is it more readable? It's fine with me, i'll change it accordingly ;)

File: src/components/AdminLink.js (hunk 2 above)
Comment (magopian) [reply src/components/AdminLink.js]: We could have, but why? What for? Is it more readable? In that case, why not

```<Link {...linkProps, to}>```?

**Comment by magopian** [comment]
@leplatrem I believe I've addressed your comments, but the [one about `linkProps`](https://github.com/Kinto/kinto-admin/pull/567#discussion_r204450727) that I might not have understood.

Please let me know what you think!

File: src/components/AdminLink.js (hunk 2 above)
Comment (leplatrem) [reply src/components/AdminLink.js]: `<Link {...linkProps, to}>` is nice. Not super important though.

**Comment by leplatrem** [comment]
> I believe I've addressed your comments, 

I still see those `const { bid, cid } = match.params;`. I think we should all write them the same way :)

**Comment by magopian** [comment]
> I still see those const { bid, cid } = match.params;. I think we should all write them the same way :)

Ah, of course, I only fixed the ones you pointed out, sorry about that, fix incoming :P

**Comment by leplatrem** [comment]
(please do not merge without squashing)

**Pull request merged**